--- a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/index.js	2018-12-01 19:42:00.645641125 +0800
+++ b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/index.js	2018-12-01 19:42:00.637642516 +0800
@@ -189,24 +189,24 @@
 			);
 
 		progressbar('#memtotal',
-			((info.memory.free + info.memory.buffered) / 1024) + ' ' + _('kB'),
-			(info.memory.total / 1024) + ' ' + _('kB'));
+			((info.memory.free + info.memory.buffered) / 1048576) + ' ' + _('MB'),
+			(info.memory.total / 1048576) + ' ' + _('MB'));
 
 		progressbar('#memfree',
-			(info.memory.free / 1024) + ' ' + _('kB'),
-			(info.memory.total / 1024) + ' ' + _('kB'));
+			(info.memory.free / 1048576) + ' ' + _('MB'),
+			(info.memory.total / 1048576) + ' ' + _('MB'));
 
 		progressbar('#membuff',
-			(info.memory.buffered / 1024) + ' ' + _('kB'),
-			(info.memory.total / 1024) + ' ' + _('kB'));
+			(info.memory.buffered / 1048576) + ' ' + _('MB'),
+			(info.memory.total / 1048576) + ' ' + _('MB'));
 
 		progressbar('#swaptotal',
-			(info.swap.free / 1024) + ' ' + _('kB'),
-			(info.swap.total / 1024) + ' ' + _('kB'));
+			(info.swap.free / 1048576) + ' ' + _('MB'),
+			(info.swap.total / 1048576) + ' ' + _('MB'));
 
 		progressbar('#swapfree',
-			(info.swap.free / 1024) + ' ' + _('kB'),
-			(info.swap.total / 1024) + ' ' + _('kB'));
+			(info.swap.free / 1048576) + ' ' + _('MB'),
+			(info.swap.total / 1048576) + ' ' + _('MB'));
 
 		progressbar('#conns',
 			info.conncount, info.connmax);
