--- a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/index.js	2019-01-23 09:53:54.398474424 +0800
+++ b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/index.js	2019-01-23 09:53:54.394474497 +0800
@@ -187,6 +187,8 @@
 				info.loadavg[1] / 65535.0,
 				info.loadavg[2] / 65535.0
 			);
+		if (e = document.getElementById('temperature'))
+			e.innerHTML = String.format('%.2f\℃' ,info.temperature / 1000);
 
 		progressbar('#memtotal',
 			((info.memory.free + info.memory.buffered) / 1024) + ' ' + _('kB'),
