--- a/package/network/services/odhcpd/files/odhcpd-update	2018-12-01 20:10:27.300212261 +0800
+++ b/package/network/services/odhcpd/files/odhcpd-update	2018-12-01 20:10:27.292212614 +0800
@@ -1,6 +1,8 @@
 #!/bin/sh
 # Make dnsmasq reread hostfile by sending SIGHUP signal
 
+logger -t odhcpd "*** ODHCPD triggers DNSMASQ reload ***"
+
 . $IPKG_INSTROOT/lib/functions/procd.sh
 
 procd_send_signal dnsmasq
--- a/package/network/services/hostapd/files/wps-hotplug.sh	2018-12-01 20:10:27.332210850 +0800
+++ b/package/network/services/hostapd/files/wps-hotplug.sh	2018-12-01 20:10:27.324211203 +0800
@@ -1,11 +1,16 @@
 #!/bin/sh
 
 if [ "$ACTION" = "pressed" -a "$BUTTON" = "wps" ]; then
+	logger "WPS button pressed, looking for active radios"
+	echo "timer" > /sys/class/leds/r7800:white:wps/trigger
 	cd /var/run/hostapd
 	for socket in *; do
 		[ -S "$socket" ] || continue
+		logger "WPS activated for: $socket"
 		hostapd_cli -i "$socket" wps_pbc
 	done
+	sleep 120
+	echo "none" > /sys/class/leds/r7800:white:wps/trigger
 fi
 
 return 0
--- a/package/network/services/hostapd/files/hostapd.sh	2018-12-01 20:10:27.364209439 +0800
+++ b/package/network/services/hostapd/files/hostapd.sh	2018-12-01 20:10:27.356209791 +0800
@@ -433,6 +433,11 @@
 		set_default wps_manufacturer "www.openwrt.org"
 		set_default wps_independent 1
 
+		# Set WPS label pin to original Netgear value stored in art partition
+		# Value for wps_pin can be overridden by /etc/config/wireless
+#		local PINdefault="$(dd if=/dev/mtdblock3 bs=1 skip=18 count=8 2>/dev/null)"
+#		set_default wps_pin "$PINdefault"
+
 		wps_state=2
 		[ -n "$wps_configured" ] && wps_state=1
 
--- a/package/network/ipv6/6in4/files/6in4.sh	2018-12-01 20:10:27.396208028 +0800
+++ b/package/network/ipv6/6in4/files/6in4.sh	2018-12-01 20:10:27.388208381 +0800
@@ -11,7 +11,7 @@
 
 proto_6in4_update() {
 	sh -c '
-		timeout=5
+		timeout=8
 
 		(while [ $((timeout--)) -gt 0 ]; do
 			sleep 1
--- a/package/base-files/files/etc/rc.button/reset	2018-12-01 20:10:27.424206793 +0800
+++ b/package/base-files/files/etc/rc.button/reset	2018-12-01 20:10:27.420206970 +0800
@@ -15,16 +15,8 @@
 	set_state failsafe
 ;;
 released)
-	if [ "$SEEN" -lt 1 ]
-	then
-		echo "REBOOT" > /dev/console
-		sync
-		reboot
-	elif [ "$SEEN" -ge 5 -a -n "$OVERLAY" ]
-	then
 		echo "FACTORY RESET" > /dev/console
 		jffs2reset -y && reboot &
-	fi
 ;;
 esac
 
--- a/target/linux/x86/64/config-default	2018-12-01 20:10:27.516202737 +0800
+++ b/target/linux/x86/64/config-default	2018-12-01 20:10:27.508203090 +0800
@@ -166,6 +166,13 @@
 CONFIG_HAVE_FENTRY=y
 # CONFIG_HAVE_GENERIC_DMA_COHERENT is not set
 CONFIG_HAVE_IRQ_EXIT_ON_IRQ_STACK=y
+CONFIG_HAVE_KVM_CPU_RELAX_INTERCEPT=y
+CONFIG_HAVE_KVM_EVENTFD=y
+CONFIG_HAVE_KVM_IRQCHIP=y
+CONFIG_HAVE_KVM_IRQFD=y
+CONFIG_HAVE_KVM_IRQ_BYPASS=y
+CONFIG_HAVE_KVM_IRQ_ROUTING=y
+CONFIG_HAVE_KVM_MSI=y
 CONFIG_HAVE_LIVEPATCH=y
 CONFIG_HAVE_MEMORY_PRESENT=y
 CONFIG_HAVE_STACK_VALIDATION=y
@@ -223,8 +230,15 @@
 # CONFIG_IOSF_MBI_DEBUG is not set
 # CONFIG_ISCSI_IBFT_FIND is not set
 CONFIG_KALLSYMS_ABSOLUTE_PERCPU=y
+CONFIG_KVM=y
+CONFIG_KVM_AMD=y
+CONFIG_KVM_ASYNC_PF=y
 # CONFIG_KVM_DEBUG_FS is not set
+CONFIG_KVM_GENERIC_DIRTYLOG_READ_PROTECT=y
 CONFIG_KVM_GUEST=y
+CONFIG_KVM_INTEL=y
+CONFIG_KVM_MMIO=y
+CONFIG_KVM_VFIO=y
 # CONFIG_LCD_CLASS_DEVICE is not set
 CONFIG_LEDS_GPIO=y
 # CONFIG_LEGACY_VSYSCALL_EMULATE is not set
--- a/feeds/packages/net/aria2/files/aria2.conf	2018-12-01 20:10:27.548201325 +0800
+++ b/feeds/packages/net/aria2/files/aria2.conf	2018-12-01 20:10:27.540201678 +0800
@@ -8,15 +8,30 @@
 # You can also add new config sections to define multi instance.
 #
 config aria2 'main'
-	option enabled '0'
 	option user 'aria2'
-	option dir '/mnt/sda1/aria2'
-	option config_dir '/var/etc/aria2'
-	option bt_enable_lpd 'true'
+	option rpc_auth_method 'none'
+	option enable_log 'false'
+	list extra_settings 'seed-ratio=0'
+	list extra_settings 'force-save=true'
+	list extra_settings 'bt-hash-check-seed=true'
+	list extra_settings 'bt-seed-unverified=true'
+	list extra_settings 'bt-save-metadata=true'
+	list extra_settings 'enable-peer-exchange=false'
+	option disk_cache '64M'
+	option user_agent 'uTorrent/3420(32126)'
+	option bt_max_peers '96'
+	option peer_id_prefix '-UT3420-'
 	option enable_dht 'true'
-	option follow_torrent 'true'
-	option file_allocation 'none'
+	option bt_enable_lpd 'true'
 	option save_session_interval '30'
+	option enabled '0'
+	option min_split_size '20M'
+	option max_concurrent_downloads '5'
+	option max_connection_per_server '16'
+	option split '5'
+	option config_dir '/etc/aria2'
+	option listen_port '36279'
+	option file_allocation 'none'
 
 	# Add addition Headers here.
 	# eg. list header 'Content-Encoding: gzip'
--- a/feeds/packages/net/aria2/files/aria2.init	2018-12-01 20:10:27.568200444 +0800
+++ b/feeds/packages/net/aria2/files/aria2.init	2018-12-01 20:10:27.564200620 +0800
@@ -152,13 +152,38 @@
 		'user_agent:string'
 }
 
+add_bttracker() {
+        gettracker=$(cat $config_file |grep "bt-tracker=")
+        if [ -z "$gettracker" ]; then
+                echo "bt-tracker=http://195.88.220.21:80/announce" >> $config_file
+        fi
+        [ ! -s "/tmp/trackers_best.txt" ] && {
+                wget -q --no-check-certificate https://raw.githubusercontent.com/ngosang/trackerslist/master/trackers_best.txt -O /tmp/trackers_best.txt
+                wget -q --no-check-certificate https://raw.githubusercontent.com/ngosang/trackerslist/master/trackers_all_https.txt -O /tmp/trackers_all_https.txt
+                [ -f "/tmp/trackers_all_https.txt" ] && cat /tmp/trackers_all_https.txt >>/tmp/trackers_best.txt && rm -f /tmp/trackers_all_https.txt
+                echo "http://tracker.skyts.net:6969/announce" >> /tmp/trackers_best.txt
+                echo "http://tracker2.itzmx.com:6961/announce" >> /tmp/trackers_best.txt
+                echo "http://tracker.skyts.cn:6969/announce" >> /tmp/trackers_best.txt
+                echo "http://tr.cili001.com:6666/announce" >> /tmp/trackers_best.txt
+        }
+        cat /tmp/trackers_best.txt |while read tracker
+        do
+                if [ -n "$tracker" ]; then
+                        newtracker=$(echo $tracker | sed 's/\//\\\//g')
+                        sed -i "s/bt-tracker=/bt-tracker=$newtracker,/" $config_file
+                fi
+        done
+	rm -f /tmp/trackers_best.txt
+}
+
 aria2_start() {
 	local section="$1"
 	aria2_validate "$section" || { _err "Validation failed."; return 1; }
 
 	[ "$enabled" = "1" ] || { _info "Instance \"${section}\" disabled."; return 1; }
-	[ -n "$dir" ] || { _err "Please set downlod dir."; return 1; }
-	[ -d "$dir" ] || { _err "Please create downlod dir first."; return 1; }
+	[ -n "$dir" ] || { _err "Please set download dir."; return 1; }
+	[ ! -d "$dir" ] && _make_dir "$dir"
+	[ -d "$dir" ] || { _err "Please create download dir first."; return 1; }
 
 	config_file="${config_dir}/${NAME}.conf.${section}"
 	config_file_tmp="${config_dir}/${NAME}.conf.tmp"
@@ -306,15 +331,17 @@
 
 	sed '/^$/d' "$config_file_tmp" >"$config_file"
 	rm -f "$config_file_tmp"
+	
+	add_bttracker
 
 	_reset_dir_mode "$config_dir"
 	_change_file_mode 600 "$config_file"
 
 	if [ -n "$user" ]; then
 		if ( user_exists "$user" && _change_owner "$user" "$config_dir" "$log" ); then
-				_info "Aria2 will run with uer '${user}'."
+				_info "Aria2 will run with user '${user}'."
 				if [ "$user" != "root" ]; then
-					_info "Please make sure user '${user}' has write access to downlod dir: ${dir}"
+					_info "Please make sure user '${user}' has write access to download dir: ${dir}"
 				fi
 		else
 			_info "Set run user to '${user}' failed, default user will be used."
--- a/feeds/packages/net/nlbwmon/files/nlbwmon.init	2018-12-01 20:10:27.592199386 +0800
+++ b/feeds/packages/net/nlbwmon/files/nlbwmon.init	2018-12-01 20:10:27.588199562 +0800
@@ -1,6 +1,6 @@
 #!/bin/sh /etc/rc.common
 
-START=60
+START=85
 
 USE_PROCD=1
 NAME=nlbwmon
@@ -59,6 +59,9 @@
 	local cfg="$1"
 	local dir
 
+	config_get enable "$cfg" enable 0
+	[ "$enable" -eq 0 ] && exit 0
+
 	config_get dir "$cfg" database_directory /var/lib/nlbwmon
 
 	mkdir -p "$dir"
--- a/feeds/packages/net/haproxy/Makefile	2018-12-01 20:10:27.616198327 +0800
+++ b/feeds/packages/net/haproxy/Makefile	2018-12-01 20:10:27.608198680 +0800
@@ -163,14 +163,8 @@
 endef
 
 define Package/haproxy/install
-	$(INSTALL_DIR) $(1)/usr/sbin
-	$(INSTALL_BIN) $(PKG_BUILD_DIR)/haproxy $(1)/usr/sbin/
-	$(INSTALL_DIR) $(1)/etc
-	$(INSTALL_CONF) ./files/haproxy.cfg $(1)/etc/
-	$(INSTALL_DIR) $(1)/etc/init.d
-	$(INSTALL_BIN) ./files/haproxy.init $(1)/etc/init.d/haproxy
-	$(INSTALL_DIR) $(1)/etc/hotplug.d/net
-	$(INSTALL_BIN) ./files/haproxy.hotplug $(1)/etc/hotplug.d/net/90-haproxy
+	$(INSTALL_DIR) $(1)/usr/bin
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/haproxy $(1)/usr/bin/
 endef
 
 Package/haproxy-nossl/install = $(Package/haproxy/install)
--- a/feeds/packages/net/nginx/Makefile	2018-12-01 20:10:27.640197269 +0800
+++ b/feeds/packages/net/nginx/Makefile	2018-12-01 20:10:27.632197621 +0800
@@ -126,6 +126,7 @@
 config_files=nginx.conf mime.types
 
 define Package/nginx/conffiles
+/etc/nginx/vhost/
 /etc/nginx/
 endef
 
@@ -348,7 +349,7 @@
 			--without-http_upstream_zone_module
 
 define Package/nginx-mod-luci/install
-	$(INSTALL_DIR) $(1)/etc/nginx
+	$(INSTALL_DIR) $(1)/etc/nginx $(1)/etc/nginx/vhost
 	$(INSTALL_BIN) ./files-luci-support/luci_uwsgi.conf $(1)/etc/nginx/luci_uwsgi.conf
 	$(INSTALL_BIN) ./files-luci-support/luci_nginx.conf $(1)/etc/nginx/luci_nginx.conf
 	$(INSTALL_DIR) $(1)/etc/uci-defaults
@@ -357,7 +358,7 @@
 
 define Package/nginx-mod-luci-ssl/install
 	$(Package/nginx-mod-luci/install)
-	$(INSTALL_DIR) $(1)/etc/nginx
+	$(INSTALL_DIR) $(1)/etc/nginx $(1)/etc/nginx/vhost
 	$(INSTALL_BIN) ./files-luci-support/luci_nginx_ssl.conf $(1)/etc/nginx/luci_nginx_ssl.conf
 	$(INSTALL_DIR) $(1)/etc/uci-defaults
 	$(INSTALL_BIN) ./files-luci-support/70_nginx-luci-support-ssl $(1)/etc/uci-defaults/70_nginx-luci-support-ssl
--- a/feeds/packages/net/nginx/files-luci-support/luci_nginx_ssl.conf	2018-12-01 20:10:27.664196210 +0800
+++ b/feeds/packages/net/nginx/files-luci-support/luci_nginx_ssl.conf	2018-12-01 20:10:27.656196563 +0800
@@ -1,64 +1,94 @@
 
-user  root;
+user nobody nogroup;
 worker_processes  1;
 
-#error_log  logs/error.log;
-#error_log  logs/error.log  notice;
-#error_log  logs/error.log  info;
+#error_log  /tmp/log/nginx/error.log;
+#error_log  /tmp/log/nginx/error.log  notice;
+#error_log  /tmp/log/nginx/error.log  info;
 
-pid        /var/run/nginx.pid;
+pid  /var/run/nginx.pid;
 
 
 events {
-    worker_connections  1024;
+	worker_connections  1024;
 }
 
 
 http {
-    include       mime.types;
-    default_type  application/octet-stream;
+	include       mime.types;
+	default_type  application/octet-stream;
 
 	sendfile on;
-    keepalive_timeout 0;
+	keepalive_timeout 0;
 	
 	client_body_buffer_size 10K;
 	client_header_buffer_size 1k;
 	client_max_body_size 1G;
 	large_client_header_buffers 2 1k;
 
-    gzip on;
-    gzip_http_version 1.1;
-    gzip_vary on;
-    gzip_comp_level 1;
-    gzip_proxied any;
+	gzip on;
+	gzip_http_version 1.1;
+	gzip_vary on;
+	gzip_comp_level 1;
+	gzip_proxied any;
+	access_log /tmp/log/nginx/access.log;
+	access_log off;
 	
 	root /www;
-	
+
 	server {
 		listen 80 default_server;
 		listen [::]:80 default_server;
-		server_name _;
-		return 301 https://$host$request_uri;
-	}
+		server_name  localhost;
+		
+		location ~ .*\.(css|js)?$ {
+			expires 15d;
+		}
+		location ~ .*\.(jpg|jpeg|gif|png|ico|woff|woff2|ttf|otf)?$ {
+			expires 365d;
+		}
+
+		location /web {
+			index  index.html;
+		}
 
-    server {
-        listen 443 ssl default_server;
+		location /webui-aria2 {
+			index  index.html;
+		}
+
+		include luci_uwsgi.conf;
+
+		#return 301 https://$host$request_uri;
+		
+	}
+	
+	server {
+		listen 443 ssl default_server;
 		listen [::]:443 ssl default_server;
-        server_name  localhost;
+		server_name  localhost;
 		
 		ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
-        ssl_prefer_server_ciphers on;
+		ssl_prefer_server_ciphers on;
 		ssl_ciphers "EECDH+ECDSA+AESGCM:EECDH+aRSA+AESGCM:EECDH+ECDSA+SHA384:EECDH+ECDSA+SHA256:EECDH+aRSA+SHA384:EECDH+aRSA+SHA256:EECDH:DHE+AESGCM:DHE:!RSA!aNULL:!eNULL:!LOW:!RC4:!3DES:!MD5:!EXP:!PSK:!SRP:!DSS:!CAMELLIA:!SEED";
-        ssl_session_tickets off;
+		ssl_session_tickets off;
 
 		ssl_certificate /etc/nginx/nginx.cer;
-        ssl_certificate_key /etc/nginx/nginx.key;
+		ssl_certificate_key /etc/nginx/nginx.key;
 		
 		location ~* .(jpg|jpeg|png|gif|ico|css|js)$ {
 			expires 365d;
 		}
 
+		location /web {
+			index  index.html;
+		}
+
+		location /webui-aria2 {
+			index  index.html;
+		}
+		
 		include luci_uwsgi.conf;
 
-    }
-}
\ No newline at end of file
+	}
+	include vhost/*.conf;
+}
--- a/feeds/packages/net/nginx/files-luci-support/luci_nginx.conf	2018-12-01 20:10:27.688195152 +0800
+++ b/feeds/packages/net/nginx/files-luci-support/luci_nginx.conf	2018-12-01 20:10:27.684195329 +0800
@@ -2,9 +2,9 @@
 user nobody nogroup;
 worker_processes  1;
 
-#error_log  logs/error.log;
-#error_log  logs/error.log  notice;
-#error_log  logs/error.log  info;
+#error_log  /tmp/log/nginx/error.log
+#error_log  /tmp/log/nginx/error.log  notice;
+#error_log  /tmp/log/nginx/error.log  info;
 
 pid        /var/run/nginx.pid;
 
@@ -46,4 +46,5 @@
 		include luci_uwsgi.conf;
 
     }
-}
\ No newline at end of file
+    include vhost/*.conf;
+}
--- a/feeds/packages/net/uwsgi-cgi/files-luci-support/uwsgi.conf	2018-12-01 20:10:27.716193918 +0800
+++ b/feeds/packages/net/uwsgi-cgi/files-luci-support/uwsgi.conf	2018-12-01 20:10:27.708194271 +0800
@@ -8,7 +8,7 @@
 buffer-size	= 10000
 reload-mercy	= 8
 max-requests	= 2000
-limit-as	= 200
+limit-as	= 1000
 reload-on-as	= 256
 reload-on-rss	= 192
 no-orphans	= true
@@ -25,3 +25,6 @@
 threads		= 3
 processes	= 3
 chmod-socket	= 666
+ignore-sigpipe=True
+ignore-write-errors=True
+disable-write-exception=True
--- a/feeds/packages/net/samba4/files/samba.init	2018-12-01 20:10:27.740192859 +0800
+++ b/feeds/packages/net/samba4/files/samba.init	2018-12-01 20:10:27.736193035 +0800
@@ -18,11 +18,19 @@
 			echo -n "${device:-$net} "
 		done
 	)
+	local name
+	config_get name                 $1 name
 
 	local workgroup description charset
 	# we dont use netbios anymore as default and wsd/avahi is dns based
 	local hostname="$(cat /proc/sys/kernel/hostname)"
-
+	
+	if [ -z "$name" ]; then
+		uci set samba4.@samba[0].name="$hostname"
+		uci commit samba4
+	else
+		hostname="$name"
+	fi
 	config_get workgroup		$1 workgroup	"WORKGROUP"
 	config_get description		$1 description	"Samba on OpenWrt"
 	config_get charset			$1 charset		"UTF-8"
@@ -31,12 +39,15 @@
 	config_get_bool DISABLE_NETBIOS	$1 disable_netbios	0
 	config_get_bool DISABLE_AD_DC	$1 disable_ad_dc	0
 	config_get_bool DISABLE_WINBIND	$1 disable_winbind	0
+	local lan_ip=$(uci -q get network.lan.ipaddr)
+	[ -z "$lan_ip" ] && lan_ip=`ifconfig br-lan 2>/dev/null | grep "inet addr:" | grep -E -o "[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+"|head -1`
+
 
 	mkdir -p /var/etc
 	sed -e "s#|NAME|#$hostname#g" \
 	    -e "s#|WORKGROUP|#$workgroup#g" \
 	    -e "s#|DESCRIPTION|#$description#g" \
-	    -e "s#|INTERFACES|#$interfaces#g" \
+	    -e "s#|INTERFACES|#$lan_ip#g" \
 	    -e "s#|CHARSET|#$charset#g" \
 	    /etc/samba/smb.conf.template > /var/etc/smb.conf
 
@@ -92,6 +103,7 @@
 	config_get_bool timemachine	$1 timemachine	0
 	config_get timemachine_maxsize $1 timemachine_maxsize
 	config_get_bool force_root	$1 force_root	0
+	config_get sharepasswd $1 sharepasswd
 
 	[ -z "$name" -o -z "$path" ] && return
 
@@ -126,6 +138,26 @@
 	fi
 	
 	[ -n "$vfs_objects" ] && echo -e "\tvfs objects = $vfs_objects" >> /var/etc/smb.conf
+        if [ -n "$sharepasswd" ] && [ -n "$users" ] && [ "$users" = "root" ]; then
+                let sharepw_num+=1
+                sharepw=$sharepasswd
+        fi
+}
+
+set_sharepw() {
+	if [ "$sharepw_num" -gt 0 ]; then
+		[ -z "$(cat /etc/samba/smb.conf |grep "#invalid users = root")" ] && \
+			sed -i 's/invalid users = root/#invalid users = root/' /etc/samba/smb.conf
+                [ -z "$(cat /etc/samba/smb.conf.template |grep "#invalid users = root")" ] && \
+                        sed -i 's/invalid users = root/#invalid users = root/' /etc/samba/smb.conf.template
+                (echo $sharepw;echo $sharepw) | smbpasswd -as root
+        else
+                [ -n "$(cat /etc/samba/smb.conf |grep "#invalid users = root")" ] && \
+                        sed -i 's/#invalid users = root/invalid users = root/' /etc/samba/smb.conf
+                [ -n "$(cat /etc/samba/smb.conf.template |grep "#invalid users = root")" ] && \
+                        sed -i 's/#invalid users = root/invalid users = root/' /etc/samba/smb.conf.template
+                [ -n "$(cat /etc/samba/smbpasswd | grep root)" ] && smbpasswd -x root
+        fi
 }
 
 init_config() {
@@ -139,9 +171,13 @@
 		chmod 0755 /var/lock
 	}
 
+	local sharepw_num=0
+	local sharepw
+
 	config_load samba4
 	config_foreach smb_header samba
 	config_foreach smb_add_share sambashare
+	set_sharepw
 }
 
 service_triggers() {
--- a/feeds/packages/net/wifischedule/net/usr/bin/wifi_schedule.sh	2018-12-01 20:10:27.768191624 +0800
+++ b/feeds/packages/net/wifischedule/net/usr/bin/wifi_schedule.sh	2018-12-01 20:10:27.760191977 +0800
@@ -305,7 +305,7 @@
 ###############################################################################
 LOGGING=$(_get_uci_value ${GLOBAL}.logging) || _exit 1
 _log ${SCRIPT} $1 $2
-lock ${LOCKFILE}
+#lock ${LOCKFILE}
 
 case "$1" in
     cron) check_cron_status ;;
--- a/feeds/packages/net/mwan3/files/etc/config/mwan3	2018-12-01 20:10:27.816189508 +0800
+++ b/feeds/packages/net/mwan3/files/etc/config/mwan3	2018-12-01 20:10:27.812189684 +0800
@@ -5,10 +5,12 @@
 
 config interface 'wan'
 	option enabled '1'
-	list track_ip '8.8.4.4'
-	list track_ip '8.8.8.8'
-	list track_ip '208.67.222.222'
-	list track_ip '208.67.220.220'
+	list track_ip 'www.qq.com'
+	list track_ip 'www.baidu.com'
+	list track_ip 'www.taobao.com'
+	list track_ip '119.29.29.29'
+	list track_ip '114.114.114.114'
+	list track_ip '1.2.4.8'
 	option family 'ipv4'
 	option reliability '2'
 	option count '1'
@@ -18,120 +20,27 @@
 	option failure_loss '20'
 	option recovery_loss '5'
 	option interval '5'
-	option down '3'
-	option up '8'
+	option down '4'
+	option up '1'
 
-config interface 'wan6'
-	option enabled '0'
-	list track_ip '2001:4860:4860::8844'
-	list track_ip '2001:4860:4860::8888'
-	list track_ip '2620:0:ccd::2'
-	list track_ip '2620:0:ccc::2'
-	option family 'ipv6'
-	option reliability '2'
-	option count '1'
-	option timeout '2'
-	option interval '5'
-	option down '3'
-	option up '8'
-
-config interface 'wanb'
-	option enabled '0'
-	list track_ip '8.8.8.8'
-	list track_ip '208.67.220.220'
-	option family 'ipv4'
-	option reliability '1'
-	option count '1'
-	option timeout '2'
-	option interval '5'
-	option down '3'
-	option up '8'
-
-config interface 'wanb6'
-	option enabled '0'
-	list track_ip '2001:4860:4860::8888'
-	list track_ip '2620:0:ccc::2'
-	option family 'ipv6'
-	option reliability '1'
-	option count '1'
-	option timeout '2'
-	option interval '5'
-	option down '3'
-	option up '8'
-
-config member 'wan_m1_w3'
-	option interface 'wan'
-	option metric '1'
-	option weight '3'
-
-config member 'wan_m2_w3'
+config member 'wan_m1_w1'
 	option interface 'wan'
-	option metric '2'
-	option weight '3'
-
-config member 'wanb_m1_w2'
-	option interface 'wanb'
-	option metric '1'
-	option weight '2'
-
-config member 'wanb_m2_w2'
-	option interface 'wanb'
-	option metric '2'
-	option weight '2'
-
-config member 'wan6_m1_w3'
-	option interface 'wan6'
 	option metric '1'
-	option weight '3'
-
-config member 'wan6_m2_w3'
-	option interface 'wan6'
-	option metric '2'
-	option weight '3'
-
-config member 'wanb6_m1_w2'
-	option interface 'wanb6'
-	option metric '1'
-	option weight '2'
-
-config member 'wanb6_m2_w2'
-	option interface 'wanb6'
-	option metric '2'
-	option weight '2'
-
-config policy 'wan_only'
-	list use_member 'wan_m1_w3'
-	list use_member 'wan6_m1_w3'
-
-config policy 'wanb_only'
-	list use_member 'wanb_m1_w2'
-	list use_member 'wanb6_m1_w2'
+	option weight '1'
 
 config policy 'balanced'
-	list use_member 'wan_m1_w3'
-	list use_member 'wanb_m1_w2'
-	list use_member 'wan6_m1_w3'
-	list use_member 'wanb6_m1_w2'
-
-config policy 'wan_wanb'
-	list use_member 'wan_m1_w3'
-	list use_member 'wanb_m2_w2'
-	list use_member 'wan6_m1_w3'
-	list use_member 'wanb6_m2_w2'
-
-config policy 'wanb_wan'
-	list use_member 'wan_m2_w3'
-	list use_member 'wanb_m1_w2'
-	list use_member 'wan6_m2_w3'
-	list use_member 'wanb6_m1_w2'
+	list use_member 'wan_m1_w1'
 
 config rule 'https'
 	option sticky '1'
 	option dest_port '443'
 	option proto 'tcp'
 	option use_policy 'balanced'
+	option enable '1'
 
 config rule 'default_rule'
 	option dest_ip '0.0.0.0/0'
 	option use_policy 'balanced'
-
+	option enable '1'
+	option sticky '0'
+	option proto 'all'
--- a/feeds/packages/net/mwan3/files/lib/mwan3/mwan3.sh	2018-12-01 20:10:27.844188273 +0800
+++ b/feeds/packages/net/mwan3/files/lib/mwan3/mwan3.sh	2018-12-01 20:10:27.840188449 +0800
@@ -860,6 +860,7 @@
 
 	rule="$1"
 
+	config_get enable $1 enable 0
 	config_get sticky $1 sticky 0
 	config_get timeout $1 timeout 600
 	config_get ipset $1 ipset
@@ -945,7 +946,7 @@
 			for IPT in "$IPT4" "$IPT6"; do
 				case $proto in
 					tcp|udp)
-					$IPT -A mwan3_rules \
+					[ "$enable" == "1" ] && $IPT -A mwan3_rules \
 						-p $proto \
 						-s $src_ip \
 						-d $dest_ip $ipset \
@@ -956,7 +957,7 @@
 						-j $policy &> /dev/null
 					;;
 					*)
-					$IPT -A mwan3_rules \
+					[ "$enable" == "1" ] && $IPT -A mwan3_rules \
 						-p $proto \
 						-s $src_ip \
 						-d $dest_ip $ipset \
@@ -971,7 +972,7 @@
 
 			case $proto in
 				tcp|udp)
-				$IPT4 -A mwan3_rules \
+				[ "$enable" == "1" ] && $IPT4 -A mwan3_rules \
 					-p $proto \
 					-s $src_ip \
 					-d $dest_ip $ipset \
@@ -982,7 +983,7 @@
 					-j $policy &> /dev/null
 				;;
 				*)
-				$IPT4 -A mwan3_rules \
+				[ "$enable" == "1" ] && $IPT4 -A mwan3_rules \
 					-p $proto \
 					-s $src_ip \
 					-d $dest_ip $ipset \
@@ -996,7 +997,7 @@
 
 			case $proto in
 				tcp|udp)
-				$IPT6 -A mwan3_rules \
+				[ "$enable" == "1" ] && $IPT6 -A mwan3_rules \
 					-p $proto \
 					-s $src_ip \
 					-d $dest_ip $ipset \
@@ -1007,7 +1008,7 @@
 					-j $policy &> /dev/null
 				;;
 				*)
-				$IPT6 -A mwan3_rules \
+				[ "$enable" == "1" ] && $IPT6 -A mwan3_rules \
 					-p $proto \
 					-s $src_ip \
 					-d $dest_ip $ipset \
--- a/feeds/packages/net/mwan3/files/usr/sbin/mwan3	2018-12-01 20:10:27.872187038 +0800
+++ b/feeds/packages/net/mwan3/files/usr/sbin/mwan3	2018-12-01 20:10:27.868187215 +0800
@@ -52,7 +52,7 @@
 		echo "Please execute \"/etc/init.d/mwan3 start\" first."
 		exit 1
 	}
-
+	
 	if [ -z "$1" ]; then
 		echo "Expecting interface. Usage: mwan3 ifup <interface>" && exit 0
 	fi
@@ -72,7 +72,7 @@
 		json_load $status
 		json_get_vars up l3_device 
 	}
-
+	
 	config_get enabled "$1" enabled 0
 
 
@@ -135,25 +135,25 @@
 	local enabled src_ip local_source
 
 	uci_toggle_state mwan3 globals enabled "1"
-
+	
 	config_get local_source globals local_source 'none'
 	[ "${local_source}" = "none" ] || {
-		src_ip=$(uci_get_state mwan3 globals src_ip)
-		[ "${src_ip}" != "" ] && {
-			ip route del default via "${src_ip}" dev lo 1>/dev/null 2>&1
-			ip addr del "${src_ip}/32" dev lo 1>/dev/null 2>&1
-		}
-
-		network_get_ipaddr src_ip "${local_source}"
-		if [ "${src_ip}" = "" ]; then
-			$LOG warn "Unable to set source ip for own initiated traffic (${local_source})"
-		else
-			ip addr add "${src_ip}/32" dev lo
-			ip route add default via "${src_ip}" dev lo
-			uci_toggle_state mwan3 globals src_ip "${src_ip}"
-		fi
+			src_ip=$(uci_get_state mwan3 globals src_ip)
+			[ "${src_ip}" != "" ] && {
+					ip route del default via "${src_ip}" dev lo 1>/dev/null 2>&1
+					ip addr del "${src_ip}/32" dev lo 1>/dev/null 2>&1
+			}
+	
+			network_get_ipaddr src_ip "${local_source}"
+			if [ "${src_ip}" = "" ]; then
+					$LOG warn "Unable to set source ip for own initiated traffic (${local_source})"
+			else
+					ip addr add "${src_ip}/32" dev lo
+					ip route add default via "${src_ip}" dev lo
+					uci_toggle_state mwan3 globals src_ip "${src_ip}"
+			fi
 	}
-
+	
 	config_foreach ifup interface
 }
 
@@ -211,11 +211,11 @@
 
 	mwan3_lock_clean
 	rm -rf $MWAN3_STATUS_DIR $MWAN3TRACK_STATUS_DIR
-
+	
 	src_ip=$(uci_get_state mwan3 globals src_ip)
 	[ "${src_ip}" = "" ] || {
-		ip route del default via "${src_ip}" dev lo 1>/dev/null 2>&1
-		ip addr del "${src_ip}/32" dev lo 1>/dev/null 2>&1
+			ip route del default via "${src_ip}" dev lo 1>/dev/null 2>&1
+			ip addr del "${src_ip}/32" dev lo 1>/dev/null 2>&1
 	}
 
 	uci_toggle_state mwan3 globals enabled "0"
--- a/feeds/luci/modules/luci-mod-system/luasrc/controller/admin/system.lua	2018-12-01 20:10:27.904185627 +0800
+++ b/feeds/luci/modules/luci-mod-system/luasrc/controller/admin/system.lua	2018-12-01 20:10:27.896185980 +0800
@@ -20,11 +20,15 @@
 		entry({"admin", "system", "admin", "sshkeys", "json"}, post_on({ keys = true }, "action_sshkeys"))
 	end
 
+	if fs.access("/usr/sbin/shellinaboxd")then
+		entry({"admin","system","WebShell"},template("admin_system/webshell"),_("WebShell"),11)
+	end
+
 	entry({"admin", "system", "startup"}, form("admin_system/startup"), _("Startup"), 45)
 	entry({"admin", "system", "crontab"}, form("admin_system/crontab"), _("Scheduled Tasks"), 46)
 
 	if fs.access("/sbin/block") and fs.access("/etc/config/fstab") then
-		entry({"admin", "system", "fstab"}, cbi("admin_system/fstab"), _("Mount Points"), 50)
+		entry({"admin", "system", "fstab"}, cbi("admin_system/fstab"), _("Mount Points / SCSI Devices"), 50)
 		entry({"admin", "system", "fstab", "mount"}, cbi("admin_system/fstab/mount"), nil).leaf = true
 		entry({"admin", "system", "fstab", "swap"},  cbi("admin_system/fstab/swap"),  nil).leaf = true
 	end
@@ -34,6 +38,7 @@
 		entry({"admin", "system", "leds"}, cbi("admin_system/leds"), _("<abbr title=\"Light Emitting Diode\">LED</abbr> Configuration"), 60)
 	end
 
+	entry({"admin", "system", "Free_Memory"},call("Free_Memory"),_("Free Memory"),65)
 	entry({"admin", "system", "flashops"}, call("action_flashops"), _("Backup / Flash Firmware"), 70)
 	entry({"admin", "system", "flashops", "reset"}, post("action_reset"))
 	entry({"admin", "system", "flashops", "backup"}, post("action_backup"))
@@ -48,6 +53,11 @@
 	entry({"admin", "system", "reboot", "call"}, post("action_reboot"))
 end
 
+function Free_Memory()
+	luci.util.exec("echo 3 > /proc/sys/vm/drop_caches")
+	luci.http.redirect(luci.dispatcher.build_url("admin","status","overview"))
+end
+
 function action_clock_status()
 	local set = tonumber(luci.http.formvalue("set"))
 	if set ~= nil and set > 0 then
--- a/feeds/luci/modules/luci-mod-network/luasrc/model/cbi/admin_network/wifi.lua	2018-12-01 20:10:27.932184393 +0800
+++ b/feeds/luci/modules/luci-mod-network/luasrc/model/cbi/admin_network/wifi.lua	2018-12-01 20:10:27.924184746 +0800
@@ -257,6 +257,16 @@
 	s:taboption("advanced", Flag, "noscan", translate("Force 40MHz mode"),
 		translate("Always use 40MHz channels even if the secondary channel overlaps. Using this option does not comply with IEEE 802.11n-2009!")).optional = true
 
+
+	log_level = s:taboption("advanced", ListValue, "log_level", translate("Logging Level"))
+	log_level.default = "2"
+	log_level.optional = true
+	log_level:value("2", translate("Informational Message"))
+	log_level:value("0", translate("Verbose Debugging"))
+	log_level:value("1", translate("Debugging"))
+	log_level:value("3", translate("Notification"))
+	log_level:value("4", translate("Warning"))
+
 	beacon_int = s:taboption("advanced", Value, "beacon_int", translate("Beacon Interval"))
 	beacon_int.optional = true
 	beacon_int.placeholder = 100
--- a/feeds/luci/modules/luci-mod-system/luasrc/view/admin_system/upgrade.htm	2018-12-01 20:10:27.984182100 +0800
+++ b/feeds/luci/modules/luci-mod-system/luasrc/view/admin_system/upgrade.htm	2018-12-01 20:10:27.980182276 +0800
@@ -39,13 +39,13 @@
 			end
 		%></li>
 		<li><% if keep then %>
-			<%:Configuration files will be kept%>
+			<%:Configuration files will be kept.%>
 		<% else %>
-			<%:Caution: Configuration files will be erased%>
+			<%:Caution: Configuration files will be erased.%>
 		<% end %></li>
 		<% if force then %>
 		<li>
-			<%:Caution: System upgrade will be forced%>
+			<%:Caution: System upgrade will be forced.%>
 		</li>
 		<% end %>
 	</ul>
--- a/feeds/luci/modules/luci-mod-system/luasrc/model/cbi/admin_system/fstab.lua	2018-12-01 20:10:28.016180689 +0800
+++ b/feeds/luci/modules/luci-mod-system/luasrc/model/cbi/admin_system/fstab.lua	2018-12-01 20:10:28.008181042 +0800
@@ -33,7 +33,7 @@
 
 block:close()
 
-m = Map("fstab", translate("Mount Points"))
+m = Map("fstab", translate("Mount Points / SCSI Devices"))
 s = m:section(TypedSection, "global", translate("Global Settings"))
 s.addremove = false
 s.anonymous = true
@@ -95,7 +95,7 @@
 
 v = m:section(Table, non_system_mounts, translate("Mounted file systems"))
 
-fs = v:option(DummyValue, "fs", translate("Filesystem"))
+f = v:option(DummyValue, "fs", translate("Filesystem"))
 
 mp = v:option(DummyValue, "mountpoint", translate("Mount Point"))
 
@@ -196,8 +196,8 @@
 	end
 end
 
-fs = mount:option(DummyValue, "fstype", translate("Filesystem"))
-fs.cfgvalue = function(self, section)
+f = mount:option(DummyValue, "fstype", translate("Filesystem"))
+f.cfgvalue = function(self, section)
 	local v, e
 
 	v = m.uci:get("fstab", section, "uuid")
@@ -269,4 +269,33 @@
 	end
 end
 
+if fs.access("/usr/bin/lsblk") and fs.access("/usr/bin/eject") then
+
+	local scsis = luci.sys.scsis()
+
+	if #scsis > 0 then
+		v = m:section(Table, scsis, translate("SCSI Devices"))
+
+		n = v:option(DummyValue, "name", translate("Name"))
+		t = v:option(DummyValue, "type", translate("Type"))
+		vr = v:option(DummyValue, "vendor", translate("Vendor"))
+		mo = v:option(DummyValue, "model", translate("Model"))
+		r = v:option(DummyValue, "rev", translate("Rev"))
+		tr = v:option(DummyValue, "tran", translate("Tran"))
+		s = v:option(DummyValue, "size", translate("Size"))
+
+		eject = v:option(Button, "eject", translate("Eject"))
+		eject.render = function(self, section, scope)
+			self.title = translate("Eject")
+			self.inputstyle = "remove"
+			Button.render(self, section, scope)
+		end
+
+		eject.write = function(self, section)
+			luci.sys.call("/usr/bin/eject -p -s /dev/'%s'" % luci.util.shellstartsqescape(scsis[section].name))
+			return luci.http.redirect(luci.dispatcher.build_url("admin/system", "fstab"))
+		end
+	end
+end
+
 return m
--- a/feeds/luci/modules/luci-base/luasrc/view/sysauth.htm	2018-12-01 20:10:28.044179455 +0800
+++ b/feeds/luci/modules/luci-base/luasrc/view/sysauth.htm	2018-12-01 20:10:28.040179631 +0800
@@ -6,7 +6,7 @@
 
 <%+header%>
 
-<form method="post" action="<%=pcdata(FULL_REQUEST_URI)%>">
+<form method="post" action="<%=pcdata(luci.http.getenv("REQUEST_URI"))%>">
 	<%- if fuser then %>
 		<div class="alert-message warning">
 			<p><%:Invalid username and/or password! Please try again.%></p>
@@ -50,6 +50,7 @@
 local fs  = require "nixio.fs"
 local https_key = uci:get("uhttpd", "main", "key")
 local https_port = uci:get("uhttpd", "main", "listen_https")
+local https_redirect = uci:get("uhttpd", "main", "redirect_https") or '0'	
 if type(https_port) == "table" then
 	https_port = https_port[1]
 end
@@ -59,7 +60,7 @@
 %>
 
 <script type="text/javascript">//<![CDATA[
-	if (document.location.protocol != 'https:') {
+	if (document.location.protocol != 'https:' && '<%=https_redirect%>' == '1') {
 		var url = 'https://' + window.location.hostname + ':' + '<%=https_port%>' + window.location.pathname;
 		var img=new Image;
 		img.onload=function(){window.location = url};
--- a/feeds/luci/modules/luci-base/luasrc/controller/admin/index.lua	2018-12-01 20:10:28.072178220 +0800
+++ b/feeds/luci/modules/luci-base/luasrc/controller/admin/index.lua	2018-12-01 20:10:28.068178396 +0800
@@ -57,6 +57,24 @@
 	page.index = true
 	toplevel_page(page, false, false)
 
+        page = node("admin", "qos")
+        page.title = _("QOS")
+        page.order = 41
+        page.index = true
+        toplevel_page(page, false, false)
+        
+        page = node("admin", "vpn")
+        page.title = _("VPN")
+        page.order = 42
+        page.index = true
+        toplevel_page(page, false, false)
+        
+        page = node("admin", "control")
+        page.title = _("Control")
+        page.order = 43
+        page.index = true
+        toplevel_page(page, false, false)
+
 	-- Even for mod-admin-full network just uses first submenu item as landing
 	page = node("admin", "network")
 	page.title = _("Network")
--- a/feeds/luci/modules/luci-base/luasrc/view/cbi/apply_widget.htm	2018-12-01 20:10:28.100176986 +0800
+++ b/feeds/luci/modules/luci-base/luasrc/view/cbi/apply_widget.htm	2018-12-01 20:10:28.096177162 +0800
@@ -122,7 +122,7 @@
 	}
 
 	function uci_apply(checked) {
-		uci_status_message('notice spinning', '<p><%:Starting configuration apply…%></p>');
+		uci_status_message('notice spinning', '<p><%:Starting configuration apply...%></p>');
 
 		xhr.post('<%=url("admin/uci")%>/' + (checked ? 'apply_rollback' : 'apply_unchecked'), uci_apply_auth, function(r, tok) {
 			if (r.status === (checked ? 200 : 204)) {
--- a/feeds/luci/modules/luci-mod-status/luasrc/view/admin_status/connections.htm	2018-12-01 20:10:28.124175928 +0800
+++ b/feeds/luci/modules/luci-mod-status/luasrc/view/admin_status/connections.htm	2018-12-01 20:10:28.120176104 +0800
@@ -131,15 +131,21 @@
 				}
 
 				label_scale.innerHTML = String.format('<%:(%d minute window, %d second interval)%>', data_wanted / 60, 3);
+				
+				var recheck_lookup_queue = {};
 
 				/* render datasets, start update interval */
 				XHR.poll(3, '<%=build_url("admin/status/realtime/connections_status")%>', null,
 					function(x, json)
 					{
-						var rows = [];
+
+						if (!json.connections)
+							return;
+
 						var conn = json.connections;
 
 						var lookup_queue = [ ];
+						var rows = [];
 
 						conn.sort(function(a, b) {
 							return b.bytes - a.bytes;
@@ -153,10 +159,10 @@
 							    (c.src == '::1'       && c.dst == '::1'))
 								continue;
 
-							if (!dns_cache[c.src])
+							if (!dns_cache[c.src] && lookup_queue.indexOf(c.src) == -1)
 								lookup_queue.push(c.src);
 
-							if (!dns_cache[c.dst])
+							if (!dns_cache[c.dst] && lookup_queue.indexOf(c.dst) == -1)
 								lookup_queue.push(c.dst);
 
 							var src = dns_cache[c.src] || (c.layer3 == 'ipv6' ? '[' + c.src + ']' : c.src);
@@ -165,22 +171,49 @@
 							rows.push([
 								c.layer3.toUpperCase(),
 								c.layer4.toUpperCase(),
-								src + ':' + c.sport,
-								dst + ':' + c.dport,
+								c.ptype,
+								[ src, c.sport ? ':' + c.sport : '' ],
+								[ dst, c.dport ? ':' + c.dport : '' ],
 								'%1024.2mB (%d <%:Pkts.%>)'.format(c.bytes, c.packets)
 							]);
 						}
 
 						cbi_update_table(conn_table, rows, '<em><%:No information available%></em>');
 
-						if (lookup_queue.length > 0)
-							XHR.get('<%=build_url("admin/status/nameinfo")%>/' + lookup_queue.slice(0, 100).join('/'), null,
-								function(x, json)
-								{
-									for (var addr in json)
-										dns_cache[addr] = json[addr];
+						if (lookup_queue.length > 0) {
+						  var reduced_lookup_queue = lookup_queue;
+						
+						  if (lookup_queue.length > 100)
+						  	reduced_lookup_queue = lookup_queue.slice(0, 100);							
+
+						  XHR.get('<%=build_url("admin/status/nameinfo")%>/' + reduced_lookup_queue.join('/'), null,
+						  	function(x, json)
+						  	{
+							  if (!json)
+								return;
+							
+							  for (var index in reduced_lookup_queue) {
+								var address = reduced_lookup_queue[index];
+								
+								if (!address)
+								  continue;
+								
+						  	    if (json[address]) {
+						  	      dns_cache[address] = json[address];
+						  	      lookup_queue.splice(reduced_lookup_queue.indexOf(address),1);
+								  continue;
 								}
-							);
+
+						  	    if(recheck_lookup_queue[address] > 2) {
+						  	      dns_cache[address] = (address.match(/:/)) ? '[' + address + ']' : address;
+						  	      lookup_queue.splice(index,1);
+						  	    } else {
+						  	      recheck_lookup_queue[address] != null ? recheck_lookup_queue[address]++ : recheck_lookup_queue[address] = 0;
+								}
+						  	  }
+						  	}
+						  );
+						}
 
 
 						var data = json.statistics;
@@ -357,6 +390,7 @@
 			<div class="tr table-titles">
 				<div class="th col-2 hide-xs"><%:Network%></div>
 				<div class="th col-2"><%:Protocol%></div>
+				<div class="th col-2"><%:Type%></div>
 				<div class="th col-7"><%:Source%></div>
 				<div class="th col-7"><%:Destination%></div>
 				<div class="th col-4"><%:Transfer%></div>
@@ -371,4 +405,4 @@
 	</div>
 </fieldset>
 
-<%+footer%>
+<%+footer%>
\ No newline at end of file
--- a/feeds/luci/modules/luci-base/luasrc/sys.lua	2018-12-01 20:10:28.176173635 +0800
+++ b/feeds/luci/modules/luci-base/luasrc/sys.lua	2018-12-01 20:10:28.168173988 +0800
@@ -70,6 +70,29 @@
 	return data
 end
 
+function scsis()
+	local data = {}
+	local ps = luci.util.execi("lsblk -d -n -P -S -o NAME,TYPE,VENDOR,MODEL,REV,TRAN,SIZE")
+ 	if not ps then
+		return
+	end
+ 	for line in ps do
+		local n, t, v, m, r, tr, s = line:match('^NAME="([^"]+)"%s+TYPE="([^"]+)"%s+VENDOR="([^"]+)"%s+MODEL="([^"]+)"%s+REV="([^"]+)"%s+TRAN="([^"]+)"%s+SIZE="([^"]+)"')
+		if n and s then
+			local row = {}
+			row.name = n
+			row.type = t
+			row.vendor = v
+			row.model = m
+			row.rev = r
+			row.tran = tr
+			row.size = s
+			table.insert(data, row)
+		end
+	end
+ 	return data
+end
+
 function mtds()
 	local data = {}
 
@@ -358,6 +381,32 @@
 					if entry[key] == nil then
 						entry[key] = val
 					end
+					if entry["ptype"] == nil then
+						val = tonumber(val)
+						if key == "sport" then
+							if (val == 443) then
+								entry["ptype"] = "Https / SSL"
+							elseif val == 80 then
+								entry["ptype"] = "Http"
+							elseif val == 53 then
+								entry["ptype"] = "Dns"
+							elseif val == 22 then
+								entry["ptype"] = "SSH"
+							elseif val == 1900 then
+								entry["ptype"] = "UPnP"
+							elseif val == 123 then
+								entry["ptype"] = "NTP"
+							elseif val == 137 then
+								entry["ptype"] = "NetBIOS"
+							elseif val == 139 or val == 445 then
+								entry["ptype"] = "Samba"	
+							elseif val == 993 then
+								entry["ptype"] = "IMAP-SSL"
+							end
+						elseif (val >= 6902) and (val <= 6968) then
+								entry["ptype"] = "Torrent"
+						end
+					end
 				elseif val then
 					entry[key] = val
 				end
--- a/feeds/luci/modules/luci-base/po/zh-cn/base.po	2018-12-01 20:10:28.208172224 +0800
+++ b/feeds/luci/modules/luci-base/po/zh-cn/base.po	2018-12-01 20:10:28.200172577 +0800
@@ -4576,7 +4576,7 @@
 msgstr "启动优先级"
 
 #: modules/luci-base/luasrc/view/cbi/apply_widget.htm:125
-msgid "Starting configuration apply…"
+msgid "Starting configuration apply..."
 msgstr "开始应用配置…"
 
 #: modules/luci-mod-network/luasrc/view/admin_network/wifi_join.htm:128
  @@ -1059,7 +1059,7 @@
 msgstr "配置失败"
 
 #: modules/luci-mod-system/luasrc/view/admin_system/upgrade.htm:42
-msgid "Configuration files will be kept"
+msgid "Configuration files will be kept."
 msgstr "将保留配置文件"
 
 #: modules/luci-base/luasrc/view/cbi/apply_widget.htm:85
--- a/feeds/luci/modules/luci-base/root/etc/config/luci	2018-12-01 20:10:28.236170990 +0800
+++ b/feeds/luci/modules/luci-base/root/etc/config/luci	2018-12-01 20:10:28.232171166 +0800
@@ -1,6 +1,6 @@
 config core main
-	option lang auto
-	option mediaurlbase /luci-static/bootstrap
+	option lang zh_cn
+	option mediaurlbase /luci-static/atmaterial
 	option resourcebase /luci-static/resources
 	
 config extern flash_keep
@@ -22,6 +22,10 @@
 	option enable 1
 		
 config internal themes
+	option Bootstrap '/luci-static/bootstrap'
+	option Material '/luci-static/material'
+	option ADVancedtomato '/luci-static/advancedtomato'
+	option ATMaterial '/luci-static/atmaterial'
 
 config internal apply
 	option rollback 30
--- a/feeds/luci/protocols/luci-proto-ppp/luasrc/model/cbi/admin_network/proto_pppoe.lua	2018-12-01 20:10:28.272169403 +0800
+++ b/feeds/luci/protocols/luci-proto-ppp/luasrc/model/cbi/admin_network/proto_pppoe.lua	2018-12-01 20:10:28.264169755 +0800
@@ -129,6 +129,8 @@
 demand.placeholder = "0"
 demand.datatype    = "uinteger"
 
+local u=net:get_interface()
+luci.tools.proto.opt_macaddr(section,u,translate("Override MAC address"))
 
 mtu = section:taboption("advanced", Value, "mtu", translate("Override MTU"))
 mtu.placeholder = "1500"
--- a/feeds/luci/applications/luci-app-wol/luasrc/model/cbi/wol.lua	2018-12-01 20:10:28.304167992 +0800
+++ b/feeds/luci/applications/luci-app-wol/luasrc/model/cbi/wol.lua	2018-12-01 20:10:28.300168168 +0800
@@ -71,7 +71,7 @@
 			local iface = luci.http.formvalue("cbid.wol.1.iface")
 			local broadcast = luci.http.formvalue("cbid.wol.1.broadcast")
 			cmd = "%s -D%s %s %q 2>&1" %{
-				util, (iface ~= "" and " -i %s" % utl.shellquote(iface) or ""),
+				util, (iface ~= "" and " -i %s" % utl.shellquote(iface) or " -i br-lan"),
 				(broadcast == "1" and " -b" or ""), mac
 			}
 		else
--- a/feeds/luci/applications/luci-app-minidlna/Makefile	2018-12-01 20:10:28.340166405 +0800
+++ b/feeds/luci/applications/luci-app-minidlna/Makefile	2018-12-01 20:10:28.336166581 +0800
@@ -7,7 +7,7 @@
 include $(TOPDIR)/rules.mk
 
 LUCI_TITLE:=LuCI Support for miniDLNA
-LUCI_DEPENDS:=+minidlna
+LUCI_DEPENDS:=+minidlna-dsd
 
 include ../../luci.mk
 
--- a/feeds/luci/applications/luci-app-ntpc/luasrc/model/cbi/ntpc/ntpc.lua	2018-12-01 20:10:28.372164994 +0800
+++ b/feeds/luci/applications/luci-app-ntpc/luasrc/model/cbi/ntpc/ntpc.lua	2018-12-01 20:10:28.368165171 +0800
@@ -23,7 +23,7 @@
 s2.addremove = false
 
 freq = s2:option(Value, "freq", translate("Offset frequency"))
-freq.datatype = "integer"
+freq.datatype = "uinteger"
 freq.rmempty = true
 
 s3 = m:section(TypedSection, "ntpserver", translate("Time Servers"))
--- a/feeds/luci/applications/luci-app-nlbwmon/luasrc/controller/nlbw.lua	2018-12-01 20:10:28.408163407 +0800
+++ b/feeds/luci/applications/luci-app-nlbwmon/luasrc/controller/nlbw.lua	2018-12-01 20:10:28.404163583 +0800
@@ -4,16 +4,16 @@
 module("luci.controller.nlbw", package.seeall)
 
 function index()
-	entry({"admin", "nlbw"}, firstchild(), _("Bandwidth Monitor"), 80)
-	entry({"admin", "nlbw", "display"}, template("nlbw/display"), _("Display"), 1)
-	entry({"admin", "nlbw", "config"}, cbi("nlbw/config"), _("Configuration"), 2)
-	entry({"admin", "nlbw", "backup"}, template("nlbw/backup"), _("Backup"), 3)
-	entry({"admin", "nlbw", "data"}, call("action_data"), nil, 4)
-	entry({"admin", "nlbw", "list"}, call("action_list"), nil, 5)
-	entry({"admin", "nlbw", "ptr"}, call("action_ptr"), nil, 6).leaf = true
-	entry({"admin", "nlbw", "download"}, call("action_download"), nil, 7)
-	entry({"admin", "nlbw", "restore"}, post("action_restore"), nil, 8)
-	entry({"admin", "nlbw", "commit"}, call("action_commit"), nil, 9)
+	entry({"admin", "status", "nlbw"}, firstchild(), _("Bandwidth Monitor"), 80)
+	entry({"admin", "status", "nlbw", "display"}, template("nlbw/display"), _("Display"), 1)
+	entry({"admin", "status", "nlbw", "config"}, cbi("nlbw/config"), _("Configuration"), 2)
+	entry({"admin", "status", "nlbw", "backup"}, template("nlbw/backup"), _("Backup"), 3)
+	entry({"admin", "status", "nlbw", "data"}, call("action_data"), nil, 4)
+	entry({"admin", "status", "nlbw", "list"}, call("action_list"), nil, 5)
+	entry({"admin", "status", "nlbw", "ptr"}, call("action_ptr"), nil, 6).leaf = true
+	entry({"admin", "status", "nlbw", "download"}, call("action_download"), nil, 7)
+	entry({"admin", "status", "nlbw", "restore"}, post("action_restore"), nil, 8)
+	entry({"admin", "status", "nlbw", "commit"}, call("action_commit"), nil, 9)
 end
 
 local function exec(cmd, args, writer)
@@ -222,6 +222,6 @@
 	local http = require "luci.http"
 	local disp = require "luci.dispatcher"
 
-	http.redirect(disp.build_url("admin/nlbw/display"))
+	http.redirect(disp.build_url("admin/status/nlbw/display"))
 	exec("/usr/sbin/nlbw", { "-c", "commit" })
 end
--- a/feeds/luci/applications/luci-app-nlbwmon/luasrc/model/cbi/nlbw/config.lua	2018-12-01 20:10:28.444161820 +0800
+++ b/feeds/luci/applications/luci-app-nlbwmon/luasrc/model/cbi/nlbw/config.lua	2018-12-01 20:10:28.440161996 +0800
@@ -12,6 +12,11 @@
 m = Map("nlbwmon", translate("Netlink Bandwidth Monitor - Configuration"),
 	translate("The Netlink Bandwidth Monitor (nlbwmon) is a lightweight, efficient traffic accounting program keeping track of bandwidth usage per host and protocol."))
 
+m.apply_on_parse=true
+function m.on_apply(self)
+luci.sys.call("/etc/init.d/nlbwmon restart> /dev/null 2>&1")
+end
+
 nw.init(luci.model.uci.cursor_state())
 
 s = m:section(TypedSection, "nlbwmon")
@@ -22,6 +27,9 @@
 s:tab("protocol", translate("Protocol Mapping"),
 	translate("Protocol mappings to distinguish traffic types per host, one mapping per line. The first value specifies the IP protocol, the second value the port number and the third column is the name of the mapped protocol."))
 
+enable=s:taboption("general",Flag,"enable",translate("启用"))
+enable.default=0
+enable.rmempty=false
 period = s:taboption("general", ListValue, "_period", translate("Accounting period"),
 	translate("Choose \"Day of month\" to restart the accounting period monthly on a specific date, e.g. every 3rd. Choose \"Fixed interval\" to restart the accounting period exactly every N days, beginning at a given date."))
 
@@ -49,7 +57,7 @@
 
 
 warning = s:taboption("general", DummyValue, "_warning", translate("Warning"))
-warning.default = translatef("Changing the accounting interval type will invalidate existing databases!<br /><strong><a href=\"%s\">Download backup</a></strong>.", luci.dispatcher.build_url("admin/nlbw/backup"))
+warning.default = translatef("Changing the accounting interval type will invalidate existing databases!<br /><strong><a href=\"%s\">Download backup</a></strong>.", luci.dispatcher.build_url("admin/status/nlbw/backup"))
 warning.rawhtml = true
 
 if (m.uci:get_first("nlbwmon", "nlbwmon", "database_interval") or ""):match("^%d%d%d%d-%d%d-%d%d/%d+$") then
@@ -127,7 +135,7 @@
 	for item in utl.imatch(ifaces:formvalue(cfg)) do
 		items[#items+1] = item
 	end
-	m:set(cfg, "local_network", items)
+	--m:set(cfg, "local_network", items)
 end
 
 
--- a/feeds/luci/applications/luci-app-nlbwmon/luasrc/view/nlbw/backup.htm	2018-12-01 20:10:28.480160233 +0800
+++ b/feeds/luci/applications/luci-app-nlbwmon/luasrc/view/nlbw/backup.htm	2018-12-01 20:10:28.472160586 +0800
@@ -11,7 +11,7 @@
 <fieldset class="cbi-section">
 	<legend><%:Restore Database Backup%></legend>
 	<p>
-		<form method="POST" action="<%=url("admin/nlbw/restore")%>" enctype="multipart/form-data">
+		<form method="POST" action="<%=url("admin/status/nlbw/restore")%>" enctype="multipart/form-data">
 			<input type="hidden" name="token" value="<%=token%>" />
 			<input type="file" name="archive" accept="application/gzip,.gz" />
 			<input type="submit" value="<%:Restore%>" class="cbi-button cbi-button-apply" />
@@ -24,7 +24,7 @@
 
 	<legend><%:Download Database Backup%></legend>
 	<p>
-		<form method="GET" action="<%=url("admin/nlbw/download")%>">
+		<form method="GET" action="<%=url("admin/status/nlbw/download")%>">
 			<input type="submit" value="<%:Generate Backup%>" class="cbi-button cbi-button-link" />
 		</form>
 	</p>
--- a/feeds/luci/applications/luci-app-nlbwmon/luasrc/view/nlbw/display.htm	2018-12-01 20:10:28.516158646 +0800
+++ b/feeds/luci/applications/luci-app-nlbwmon/luasrc/view/nlbw/display.htm	2018-12-01 20:10:28.512158822 +0800
@@ -394,7 +394,7 @@
 
 function fetchData(period)
 {
-	XHR.get('<%=url("admin/nlbw/data")%>', { period: period, group_by: 'family,mac,ip,layer7', order_by: '-rx_bytes,-tx_bytes' }, function(xhr, res) {
+	XHR.get('<%=url("admin/status/nlbw/data")%>', { period: period, group_by: 'family,mac,ip,layer7', order_by: '-rx_bytes,-tx_bytes' }, function(xhr, res) {
 		if (res !== null && typeof(res) === 'object' && typeof(res.columns) === 'object' && typeof(res.data) === 'object')
 			trafficData = res;
 
@@ -409,7 +409,7 @@
 		renderLayer7Data();
 		renderIPv6Data();
 
-		XHR.get('<%=url("admin/nlbw/ptr")%>/' + ipAddrs.join('/'), null, function(xhr, res) {
+		XHR.get('<%=url("admin/status/nlbw/ptr")%>/' + ipAddrs.join('/'), null, function(xhr, res) {
 			if (res !== null && typeof(res) === 'object')
 				hostNames = res;
 		});
@@ -633,7 +633,7 @@
 		var cell = table.insertRow(-1).insertCell(-1);
 
 		cell.setAttribute('colspan', 6);
-		cell.innerHTML = '<em><%:No data recorded yet.%> <a href="<%=url("admin/nlbw/commit")%>"><%:Force reload…%></a></em>';
+		cell.innerHTML = '<em><%:No data recorded yet.%> <a href="<%=url("admin/status/nlbw/commit")%>"><%:Force reload…%></a></em>';
 	}
 
 	pie('traf-pie', trafData);
@@ -693,7 +693,7 @@
 		var cell = table.insertRow(-1).insertCell(-1);
 
 		cell.setAttribute('colspan', 6);
-		cell.innerHTML = '<em><%:No data recorded yet.%> <a href="<%=url("admin/nlbw/commit")%>"><%:Force reload…%></a></em>';
+		cell.innerHTML = '<em><%:No data recorded yet.%> <a href="<%=url("admin/status/nlbw/commit")%>"><%:Force reload…%></a></em>';
 	}
 
 	pie('layer7-rx-pie', rxData);
@@ -811,7 +811,7 @@
 		var cell = table.insertRow(-1).insertCell(-1);
 
 		cell.setAttribute('colspan', 7);
-		cell.innerHTML = '<em><%:No data recorded yet.%> <a href="<%=url("admin/nlbw/commit")%>"><%:Force reload…%></a></em>';
+		cell.innerHTML = '<em><%:No data recorded yet.%> <a href="<%=url("admin/status/nlbw/commit")%>"><%:Force reload…%></a></em>';
 	}
 
 	var shareData = [], hostsData = [];
@@ -1012,10 +1012,10 @@
 
 <div class="cbi-section" id="container.nlbw.export" style="display:none">
 	<ul>
-		<li><a href="<%=url('admin/nlbw/data')%>?type=csv&#38;group_by=mac&#38;order_by=-rx,-tx"><%:CSV, grouped by MAC%></a></li>
-		<li><a href="<%=url('admin/nlbw/data')%>?type=csv&#38;group_by=ip&#38;order_by=-rx,-tx"><%:CSV, grouped by IP%></a></li>
-		<li><a href="<%=url('admin/nlbw/data')%>?type=csv&#38;group_by=layer7&#38;order_by=-rx,-tx"><%:CSV, grouped by protocol%></a></li>
-		<li><a href="<%=url('admin/nlbw/data')%>?type=json"><%:JSON dump%></a></li>
+		<li><a href="<%=url('admin/status/nlbw/data')%>?type=csv&#38;group_by=mac&#38;order_by=-rx,-tx"><%:CSV, grouped by MAC%></a></li>
+		<li><a href="<%=url('admin/status/nlbw/data')%>?type=csv&#38;group_by=ip&#38;order_by=-rx,-tx"><%:CSV, grouped by IP%></a></li>
+		<li><a href="<%=url('admin/status/nlbw/data')%>?type=csv&#38;group_by=layer7&#38;order_by=-rx,-tx"><%:CSV, grouped by protocol%></a></li>
+		<li><a href="<%=url('admin/status/nlbw/data')%>?type=json"><%:JSON dump%></a></li>
 	</ul>
 </div>
 
@@ -1025,7 +1025,7 @@
 	cbi_t_add('nlbw', 'ipv6');
 	cbi_t_add('nlbw', 'export');
 
-	XHR.get('<%=url("admin/nlbw/list")%>', null, function(xhr, res) {
+	XHR.get('<%=url("admin/status/nlbw/list")%>', null, function(xhr, res) {
 
 		if (res !== null && typeof(res) === 'object' && res.length > 0) {
 			trafficPeriods = res;
--- a/feeds/luci/applications/luci-app-wifischedule/luasrc/model/cbi/wifischedule/wifi_schedule.lua	2018-12-01 20:10:28.552157059 +0800
+++ b/feeds/luci/applications/luci-app-wifischedule/luasrc/model/cbi/wifischedule/wifi_schedule.lua	2018-12-01 20:10:28.548157235 +0800
@@ -37,7 +37,8 @@
 
 -- BEGIN Map
 m = Map("wifi_schedule", translate("Wifi Schedule"), translate("Defines a schedule when to turn on and off wifi."))
-function m.on_commit(self)
+m.apply_on_parse=true
+function m.on_apply(self)
     sys.exec("/usr/bin/wifi_schedule.sh cron")
 end
 -- END Map
--- a/feeds/luci/applications/luci-app-samba4/luasrc/model/cbi/samba4.lua	2018-12-01 20:10:28.592155296 +0800
+++ b/feeds/luci/applications/luci-app-samba4/luasrc/model/cbi/samba4.lua	2018-12-01 20:10:28.588155472 +0800
@@ -75,6 +75,9 @@
 au = s:option(Value, "users", translate("Allowed users"))
 au.rmempty = true
 
+sp = s:option(Value, "sharepasswd", translate("Share Password"))
+sp.rmempty = true
+
 go = s:option(Flag, "guest_ok", translate("Allow guests"))
 go.enabled = "yes"
 go.disabled = "no"
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/controller/mwan3.lua	2018-12-01 20:10:28.704150357 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/controller/mwan3.lua	2018-12-01 20:10:28.700150534 +0800
@@ -125,7 +125,7 @@
 		if not dump then
 			dump = require("luci.util").ubus("network.interface.%s" % interface, "status", {})
 		end
-
+		
 		if dump and dump.route then
 			local _, route
 			for _, route in ipairs(dump.route) do
@@ -160,7 +160,7 @@
 			end
 		elseif task == "ping_trackips" then
 			local trackips = uci:get("mwan3", interface, "track_ip")
-			if #trackips > 0 then
+			if trackips and #trackips > 0 then
 				for i in pairs(trackips) do
 					diag_command("ping -I %s -c 5 -W 1 %s 2>&1", device, trackips[i])
 				end
@@ -193,11 +193,11 @@
 			local number = getInterfaceNumber(interface)
 			local routeTable = sys.exec(string.format("ip route list table %s", number))
 			if routeTable ~= "" then
-				luci.http.write(i18n.translatef("Routing table %s for interface %s found", number, interface))
+				luci.http.write(i18n.translatef("Routing table %s for interface %s found", interface, number))
 				luci.http.write("\n")
 				luci.http.write(routeTable)
 			else
-				luci.http.write(i18n.translatef("Routing table %s for interface %s not found", number, interface))
+				luci.http.write(i18n.translatef("Routing table %s for interface %s not found", interface, number))
 			end
 		elseif task == "hotplug_ifup" then
 			os.execute(string.format("/usr/sbin/mwan3 ifup %s", ut.shellquote(interface)))
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/interface.lua	2018-12-01 20:10:28.744148594 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/interface.lua	2018-12-01 20:10:28.736148947 +0800
@@ -8,38 +8,39 @@
 
 function interfaceWarnings(overview, count, iface_max)
 	local warnings = ""
+	local s_format = "<font color=\"ff0000\"><strong>%s</strong></font></br>"
 	if count <= iface_max then
-		warnings = string.format("<strong>%s</strong><br />",
+		warnings = string.format("<strong>%s</strong></br>",
 			translatef("There are currently %d of %d supported interfaces configured", count, iface_max)
 			)
 	else
-		warnings = string.format("<strong>%s</strong><br />",
+		warnings = string.format(s_format,
 			translatef("WARNING: %d interfaces are configured exceeding the maximum of %d!", count, iface_max)
 			)
 	end
 
 	for i, k in pairs(overview) do
 		if overview[i]["network"] == false then
-			warnings = warnings .. string.format("<strong>%s</strong><br />",
+			warnings = warnings .. string.format(s_format,
 					translatef("WARNING: Interface %s are not found in /etc/config/network", i)
 					)
 		end
 
 		if overview[i]["default_route"] == false then
-			warnings = warnings .. string.format("<strong>%s</strong><br />",
+			warnings = warnings .. string.format(s_format,
 				translatef("WARNING: Interface %s has no default route in the main routing table", i)
 				)
 		end
 
 		if overview[i]["reliability"] == false then
-			warnings = warnings .. string.format("<strong>%s</strong><br />",
+			warnings = warnings .. string.format(s_format,
 				translatef("WARNING: Interface %s has a higher reliability " ..
 				"requirement than tracking hosts (%d)", i, overview[i]["tracking"])
 				)
 		end
 
 		if overview[i]["duplicate_metric"] == true then
-			warnings = warnings .. string.format("<strong>%s</strong><br />",
+			warnings = warnings .. string.format(s_format,
 				translatef("WARNING: Interface %s has a duplicate metric %s configured", i, overview[i]["metric"])
 				)
 		end
@@ -104,7 +105,7 @@
 			end
 		end
 	)
-
+	
 	-- calculate iface_max usage from firewall mmx_mask
 	function bit(p)
 		return 2 ^ (p - 1)
@@ -140,7 +141,7 @@
 mwan_interface = m5:section(TypedSection, "interface", nil,
 	translate("MWAN supports up to 252 physical and/or logical interfaces<br />" ..
 	"MWAN requires that all interfaces have a unique metric configured in /etc/config/network<br />" ..
-	"Names must match the interface name found in /etc/config/network<br />" ..
+	"Names must match the interface name found in /etc/config/network (see advanced tab)<br />" ..
 	"Names may contain characters A-Z, a-z, 0-9, _ and no spaces<br />" ..
 	"Interfaces may not share the same name as configured members, policies or rules"))
 mwan_interface.addremove = true
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/policy.lua	2018-12-01 20:10:28.784146831 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/policy.lua	2018-12-01 20:10:28.776147183 +0800
@@ -25,7 +25,8 @@
 	local warnings = ""
 	for i, k in pairs(policy_error) do
 		if policy_error[i] == true then
-			warnings = warnings .. string.format("<strong>%s</strong><br />",
+			local s_format = "<font color=\"ff0000\"><strong>%s</strong></font></br>"
+			warnings = warnings .. string.format(s_format,
 				translatef("WARNING: Policy %s has exceeding the maximum name of 15 characters", i)
 				)
 		end
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/rule.lua	2018-12-01 20:10:28.820145244 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/rule.lua	2018-12-01 20:10:28.816145420 +0800
@@ -29,7 +29,8 @@
 	local warnings = ""
 	for i, k in pairs(rule_error) do
 		if rule_error[i] == true then
-			warnings = warnings .. string.format("<strong>%s</strong><br />",
+			local s_format = "<font color=\"ff0000\"><strong>%s</strong></font></br>"
+			warnings = warnings .. string.format(s_format,
 				translatef("WARNING: Rule %s have a port configured with no or improper protocol specified!", i)
 				)
 		end
@@ -65,6 +66,9 @@
 	luci.http.redirect(dsp.build_url("admin", "network", "mwan", "rule", section))
 end
 
+enable = mwan_rule:option(Flag,"enable",translate("enable"))
+	enable.rmempty = false
+
 src_ip = mwan_rule:option(DummyValue, "src_ip", translate("Source address"))
 src_ip.rawhtml = true
 function src_ip.cfgvalue(self, s)
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/ruleconfig.lua	2018-12-01 20:10:28.860143480 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/model/cbi/mwan/ruleconfig.lua	2018-12-01 20:10:28.856143657 +0800
@@ -13,6 +13,9 @@
 mwan_rule.addremove = false
 mwan_rule.dynamic = false
 
+enable = mwan_rule:option(Flag,"enable",translate("enable"))
+	enable.rmempty = false
+
 src_ip = mwan_rule:option(Value, "src_ip", translate("Source address"),
 	translate("Supports CIDR notation (eg \"192.168.100.0/24\") without quotes"))
 src_ip.datatype = ipaddr
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/overview_status_interface.htm	2018-12-01 20:10:28.940139953 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/overview_status_interface.htm	2018-12-01 20:10:28.932140306 +0800
@@ -15,7 +15,7 @@
 				for ( var iface in status.interfaces)
 				{
 					var state = '';
-					var css = ''
+					var css = '';
 					switch (status.interfaces[iface].status)
 					{
 						case 'online':
@@ -36,11 +36,11 @@
 						css
 					);
 					statusview += String.format(
-						'<div><strong>Interface: </strong>%s</div>',
+						'<div><strong><%:Interface%>: </strong>%s</div>',
 						iface
 					);
 					statusview += String.format(
-						'<div><strong>Status: </strong>%s</div>',
+						'<div><strong><%:Status%>: </strong>%s</div>',
 						state
 					);
 					statusview += '</div>'
@@ -64,10 +64,19 @@
 		float: left;
 		line-height: 125%;
 	}
+	.success {
+		background-color: rgb(144, 240, 144);
+	}
+	.danger {
+		background-color: rgb(240, 144, 144);
+	}
+	.warning {
+		background-color: rgb(255, 229, 153);
+	}
 </style>
 
 <fieldset id="interface_field" class="cbi-section">
-	<legend><%:MWAN Interfaces%></legend>
+	<legend><%:MWAN Interfaces - Live Status%></legend>
 	<div id="mwan_status_text">
 		<img src="<%=resource%>/icons/loading.gif" alt="<%:Loading%>" style="vertical-align:middle" />
 		<%:Collecting data...%>
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_detail.htm	2018-12-01 20:10:28.976138366 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_detail.htm	2018-12-01 20:10:28.972138543 +0800
@@ -13,6 +13,7 @@
 	<li class="cbi-tab-disabled"><a href="<%=luci.dispatcher.build_url("admin/status/mwan/troubleshooting")%>"><%:Troubleshooting%></a></li>
 </ul>
 
+<script type="text/javascript" src="<%=resource%>/cbi.js"></script>
 <script type="text/javascript">//<![CDATA[
 	XHR.poll(5, '<%=luci.dispatcher.build_url("admin", "status", "mwan", "detailed_status")%>', null,
 		function(x)
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_diagnostics.htm	2018-12-01 20:10:29.016136603 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_diagnostics.htm	2018-12-01 20:10:29.012136779 +0800
@@ -25,6 +25,7 @@
 	)
 %>
 
+<script type="text/javascript" src="<%=resource%>/cbi.js"></script>
 <script type="text/javascript">//<![CDATA[
 	var stxhr = new XHR();
 
@@ -59,14 +60,14 @@
 			<br />
 
 			<div style="width:30%; float:left">
-				<label class="cbi-value-title"><%:Interface%></label>
+				<label class="cbi-value-title" style="width:auto"><%:Interface%></label>
 				<select name="iface" style="width:auto">
 					<% for _, z in ipairs(iface) do -%><option value="<%=z%>"><%=z%></option><%- end %>
 				</select>
 			</div>
 
 			<div style="width:30%; float:left">
-				<label class="cbi-value-title"><%:Task%></label>
+				<label class="cbi-value-title" style="width:auto"><%:Task%></label>
 				<select name="task" style="width:auto">
 					<option value="ping_gateway"><%:Ping default gateway%></option>
 					<option value="ping_trackips"><%:Ping tracking IP%></option>
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_interface.htm	2018-12-01 20:10:29.060134663 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_interface.htm	2018-12-01 20:10:29.052135016 +0800
@@ -13,6 +13,7 @@
 	<li class="cbi-tab-disabled"><a href="<%=luci.dispatcher.build_url("admin/status/mwan/troubleshooting")%>"><%:Troubleshooting%></a></li>
 </ul>
 
+<script type="text/javascript" src="<%=resource%>/cbi.js"></script>
 
 <div class="cbi-map">
 	<%+mwan/overview_status_interface%>
--- a/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_troubleshooting.htm	2018-12-01 20:10:29.088133429 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/luasrc/view/mwan/status_troubleshooting.htm	2018-12-01 20:10:29.084133605 +0800
@@ -13,6 +13,7 @@
 	<li class="cbi-tab"><a href="<%=luci.dispatcher.build_url("admin/status/mwan/troubleshooting")%>"><%:Troubleshooting%></a></li>
 </ul>
 
+<script type="text/javascript" src="<%=resource%>/cbi.js"></script>
 <script type="text/javascript">//<![CDATA[
 	XHR.poll(15, '<%=luci.dispatcher.build_url("admin", "status", "mwan", "troubleshooting_display")%>', null,
 		function(x)
--- a/feeds/luci/applications/luci-app-mwan3/po/zh-cn/mwan3.po	2018-12-01 20:10:29.124131843 +0800
+++ b/feeds/luci/applications/luci-app-mwan3/po/zh-cn/mwan3.po	2018-12-01 20:10:29.116132195 +0800
@@ -325,6 +325,9 @@
 msgstr "MWAN 规则配置 - %s"
 
 #: applications/luci-app-mwan3/luasrc/view/mwan/status_detail.htm:27
+msgid "MWAN Interfaces - Live Status"
+msgstr "MWAN 状态 - 接口实时状态"
+
 msgid "MWAN Status - Detail"
 msgstr "MWAN 状态 - 详细"
 
@@ -338,12 +341,11 @@
 
 #: applications/luci-app-mwan3/luasrc/model/cbi/mwan/interface.lua:141
 msgid ""
-"MWAN supports up to 252 physical and/or logical interfaces<br />MWAN "
-"requires that all interfaces have a unique metric configured in /etc/config/"
-"network<br />Names must match the interface name found in /etc/config/"
-"network<br />Names may contain characters A-Z, a-z, 0-9, _ and no spaces<br /"
-">Interfaces may not share the same name as configured members, policies or "
-"rules"
+"MWAN supports up to 252 physical and/or logical interfaces<br />"
+"MWAN requires that all interfaces have a unique metric configured in /etc/config/network<br />"
+"Names must match the interface name found in /etc/config/network (see advanced tab)<br />"
+"Names may contain characters A-Z, a-z, 0-9, _ and no spaces<br />"
+"Interfaces may not share the same name as configured members, policies or rules"
 msgstr ""
 "MWAN 支持最多 252 个物理或逻辑接口。<br />MWAN 要求所有接口必须在 /etc/"
 "config/network 中设定唯一的网关跃点。<br />名称必须与 /etc/config/network 中"
@@ -539,11 +541,11 @@
 
 #: applications/luci-app-mwan3/luasrc/controller/mwan3.lua:196
 msgid "Routing table %s for interface %s found"
-msgstr "找到路由表 %s，为接口 %s"
+msgstr "找到接口 %s 所在路由表 %s 的信息"
 
 #: applications/luci-app-mwan3/luasrc/controller/mwan3.lua:200
 msgid "Routing table %s for interface %s not found"
-msgstr "没有找到路由表 %s，为接口 %s"
+msgstr "未找到接口 %s 所在路由表 %s 的信息"
 
 #: applications/luci-app-mwan3/luasrc/model/cbi/mwan/globalsconfig.lua:35
 msgid "Routing table lookup"
--- a/feeds/luci/applications/luci-app-nft-qos/luasrc/controller/nft-qos.lua	2018-12-01 20:10:29.152130608 +0800
+++ b/feeds/luci/applications/luci-app-nft-qos/luasrc/controller/nft-qos.lua	2018-12-01 20:10:29.148130785 +0800
@@ -12,7 +12,7 @@
 		template("nft-qos/rate"), _("Rate"), 5).leaf = true
 	entry({"admin", "status", "realtime", "rate_status"},
 		call("action_rate")).leaf = true
-	entry({"admin", "services", "nft-qos"}, cbi("nft-qos/nft-qos"),
+	entry({"admin", "qos", "nft-qos"}, cbi("nft-qos/nft-qos"),
 		_("Qos over Nftables"), 60)
 end
 
